### PR TITLE
Add `quote_identifiers` parameter to `unpivot` to handle case-sensitive column names

### LINKED ISSUE #216: Enable quoting column names for `unpivot` macro
I have a table whose column names are quoted, like "BX 759 - long column name".

When I try to unpivot this table, the column names fetched don't contain quotes and the macro doesn't put the quotes in them, thus the macro fails.

### PR DESCRIPTION
**Pull request by error418**
resolves #216

## Description & motivation
Added `quote_identifiers` parameter to `unpivot` to handle case-sensitive column names.

This is like https://github.com/dbt-labs/dbt-utils/pull/135 was for `pivot`.

## Checklist
- [x] This code is associated with an Issue which has been triaged and [accepted for development](https://docs.getdbt.com/docs/contributing/oss-expectations#pull-requests). 
- [x] I have verified that these changes work locally on the following warehouses (Note: it's okay if you do not have access to all warehouses, this helps us understand what has been covered)
    - [x] BigQuery (tested by @Ishankoradia - see PR comments)
    - [x] Postgres
    - [ ] Redshift
    - [ ] Snowflake
- [x] I followed guidelines to ensure that my changes will work on "non-core" adapters by:
    - [x] dispatching any new macro(s) so non-core adapters can also use them (e.g. [the `star()` source](https://github.com/dbt-labs/dbt-utils/blob/main/macros/sql/star.sql))
    - [x] using the `limit_zero()` macro in place of the literal string: `limit 0`
    - [x] using `dbt.type_*` macros instead of explicit datatypes (e.g. `dbt.type_timestamp()` instead of `TIMESTAMP`
- [x] I have updated the README.md (if applicable)
- [x] I have added tests & descriptions to my models (and macros if applicable)
- [x] I have added an entry to CHANGELOG.md

### CODE REVIEW

**Comment by error418** [comment]
CircleCI Jobs for BigQuery, Redshift and Snowflake fail due to database connection errors. This needs to be fixed in the Build server configuration.

File: integration_tests/models/sql/test_unpivot_quote.sql
Comment (joellabes) [inline integration_tests/models/sql/test_unpivot_quote.sql:5]: Does lowercasing break this test, or just make it redundant? 

I assume the former, in which case `enabled: "{{ target.name == 'redshift' }}"` is tidier on the test in your schema.yml

File: README.md
Comment (joellabes) [inline README.md:1213]:
```suggestion
- `quote_identifiers` (optional, default=`False`): will encase selected columns and aliases in quotes according to your adapter's implementation of `adapter.quote` (e.g. `"field_name" as "field_name"`).
```

File: integration_tests/models/sql/test_unpivot_quote.sql
Comment (joellabes) [inline integration_tests/models/sql/test_unpivot_quote.sql:12]: Do these actually need to be quoted?

**Review by joellabes — DISMISSED** [review]
Thanks @error418! The changes to the macro look good; I have a few nitpicks around the tests though ⛏️

File: integration_tests/models/sql/test_unpivot_quote.sql
Comment (error418) [reply integration_tests/models/sql/test_unpivot_quote.sql:5]: Nice - this is way cleaner than the workaround I came up with

File: integration_tests/models/sql/test_unpivot_quote.sql
Comment (error418) [reply integration_tests/models/sql/test_unpivot_quote.sql:12]: BigQuery seems to have problems with unquoted columns since it uppercases all columns - I will doublecheck this to be sure

Comment (error418) [reply integration_tests/models/sql/test_unpivot_quote.sql:6]: Quoting the column names is required for snowflake. Other tests handle this conditionally:

https://github.com/dbt-labs/dbt-utils/blob/965726ea4a1cf2d793ec93ade5d39a884c8dc577/integration_tests/models/sql/test_unpivot_bool.sql#L8-L12

Comment (error418) [reply integration_tests/models/sql/test_unpivot_quote.sql:6]: @joellabes I added a comment mentioning the cause for the quotes in the test itself. Is this sufficient?

**Comment by error418** [comment]
rebased to current main and resolved merge conflicts in `CHANGELOG.md`

**Comment by error418** [comment]
Anything missing to get this merged?

**Comment by error418** [comment]
ping

**Comment by Shadowsong27** [comment]
Any upates on this? I have encountered a similar scenario where i use date `yyyy-mm-dd` as the format, it end up evaluating  the date string into an arithmic operation. Quoting seems quite useful here.

**Comment by Ishankoradia** [comment]
Hi @joellabes and @error418 what are we waiting on for this to be merged ?
I have faced a similar issue

**Comment by Ishankoradia** [comment]
@error418 and @joellabes  i have tested the changes for bigquery warehouse also and its work fine.

**Comment by Abhishek-N** [comment]
hey @error418 @joellabes, facing a similar issue and need to have this merged. can you merge this asap please?

**Comment by error418** [comment]
@Shadowsong27 @Ishankoradia @Abhishek-N at the moment I am waiting for a review by @joellabes (or any maintainer of `dbt-utils`).

@Ishankoradia thank you for testing the changes for BigQuery

Should there be any remaining issues with this PR I am happy to provide the necessary fixes.

**Comment by pnd-mnicolau** [comment]
bump

**Comment by alejobs** [comment]
What's missing here? I need this fix to be deployed. Thanks.

**Comment by MiguelMadero** [comment]
What's missing here? I need this fix to be deployed. Thanks.

**Comment by error418** [comment]
@joellabes could you please provide some information about the status of this pull request?

This is stalling for a while now without any feedback from the maintainers, which sadly becomes a bit frustrating.

Thanks

**Comment by dbeatty10** [comment]
Hi all,

This is our highest priority PR to review right now, and I'll be diving deeper into it this week.

After taking an initial look, the heart of the implementation looks good.

The main thing I'd like to follow-up on prior to approving + merging is the testing:
- It looks like there is special configuration in the CI tests so that Snowflake and Redshift pass

So I'd like to:
- Make sure that the test is running for Redshift.
- Align how Snowflake is handled across similar tests. i.e., either standardize on [prior art](https://github.com/dbt-labs/dbt-utils/blob/965726ea4a1cf2d793ec93ade5d39a884c8dc577/integration_tests/models/sql/test_unpivot_bool.sql#L8-L12) or standardize on the [`adapter.quote()` approach](https://github.com/dbt-labs/dbt-utils/pull/792/files#r1194616888) in this PR

**Comment by dbeatty10** [comment]
@error418 I think I figured out the root cause of the Redshift test failures in CI.

As you noted, the _default_ behavior in Redshift is to lowercase all column names _even if_ a table is created with quoted mixed-case column names.

It looks like the solution is to set [`enable_case_sensitive_identifier` to `true`](https://docs.aws.amazon.com/redshift/latest/dg/r_enable_case_sensitive_identifier.html#w265aac71c45b7). I'm going to try  adding that configuration to the CI cluster or workgroup's parameter group and see if that allows this test to pass for Redshift.

File: macros/sql/unpivot.sql
Comment (dbeatty10) [inline macros/sql/unpivot.sql:15]: In an ideal world, I would rather have the default be `quote_identifiers=True` so that it matches with `pivot`. But agree with your choice here to default it to `False` instead so we don't have any breaking behavior change.

**Review by dbeatty10 — APPROVED** [review]
We're ready to rock @error418 🎸 

The tests in Redshift are passing after setting `enable_case_sensitive_identifier` to `true` in the [cluster’s parameter group](https://docs.aws.amazon.com/redshift/latest/mgmt/managing-parameter-groups-console.html#parameter-group-modify) via the [Amazon Redshift console](https://docs.aws.amazon.com/redshift/latest/mgmt/managing-parameter-groups-console.html#parameter-group-modify).

**Pull request merged**
